### ISSUE TITLE
eBPF data plane fails to forward UDP packet fragments

### ISSUE THREAD

**Issue by nick-oconnor** [projectcalico/calico#8821]
## Expected Behavior

UDP packet fragments destined for a pod's IP which are not denied by policy arrive on the pod's interface.

## Current Behavior

The eBPF data plane appears to be dropping UDP packet fragments by policy. The initial fragment is correctly forwarded from the node interface to the pod interface, but subsequent fragments do not appear on the pod's interface. When a UDP packet fragment is dropped, calico's dropped by policy counter for the interface is incremented. The pod interface eventually responds with "fragment reassembly time exceeded".

The only policies I have defined are k8s network policies and this problem does not occur when using the IPTables data plane. 

## Possible Solution

No idea. There may be a bug in calico's eBPF code.

## Steps to Reproduce (for bugs)

1. Enable the eBPF data plane (with or without DSR, I have kube-proxy scaled to zero)
2. Deploy a pod
3. Start a packet capture on the node running the pod
4. Send a fragmented UDP packet to the pod IP (I'm unsure how to replicate this outside of SNMP)

## Context

I experienced this behavior after migrating from the IPTables data plane to the eBPF data plane. All SNMP responses exceeding the network's MTU caused my SNMP collector to timeout.

## Your Environment

* Calico version: v3.28.0
* Orchestrator version (e.g. kubernetes, mesos, rkt): 1.29.4
* Operating System and version: Ubuntu 24.04 (6.8.0-31 kernel)

**Comment by tomastigera** [comment]
That is correct observation. Unfortunately, ebpf dataplane does not support ip fragmentation as only the first fragment contains udp ports. The subsequent fragments cannot be matched reliably with the ongoing flow. We cannot reassemble the fragments in eBPF easily (that is a limitation of the technology). This said, we might consider some improvements/workarounds in a future release.

**Comment by nick-oconnor** [comment]
@tomastigera Wow thanks for the quick reply! Very interesting. Looks like I have some homework regarding eBPF APIs. It'll probably save folks some time by adding this to the eBPF docs for Calico.

**Comment by nick-oconnor** [comment]
Related: https://github.com/cilium/cilium/issues/25709#issuecomment-2105977944

**Comment by tomastigera** [comment]
Thanks for the pointer. Problem with kfunc is that they are in "newer" kernels only and are not necessarily a stable API. But  we could perhaps add it for kernels that have that feature! :+1:

**Comment by tomastigera** [comment]
Seems like the patch :arrow_up: is not present in any released kernel :(

**Comment by diogenxs** [comment]
also facing this issue, in my case I noticed that the error only happens when the target is a service IP, if I test from a pod to pod IP it works, would that make sense?

**Comment by tomastigera** [comment]
@diogenxs do you have a different MTU on the pod-pod path than on the "default" route as that is probably what decides the MTu for the service path (larger) ? Do you use overlay (vxlan) ? What is the MTU on your devices?

**Comment by diogenxs** [comment]
> do you have a different MTU on the pod-pod path than on the "default" route

no, both routes pod and services, goes to the same path

> Do you use overlay (vxlan)?

no, `encapsulation: None` set on `Installation`  of `tigera-operator`

> What is the MTU on your devices?

1472, I tried to force Calico to use a lower one but didn't have any lucky
 

I'm advertising the [services IPs trough BGP](https://docs.tigera.io/calico/latest/networking/configuring/advertise-service-ips#advertise-service-cluster-ip-addresses), same as pods, one thing I noticed is that some services have a "static" entry at node level, what is the desired state here? does every service IP need to be a static entry as well?
```
# ip r  | grep 10.226.2
10.226.2.10 via 169.254.1.1 dev bpfin.cali 
10.226.2.20 via 169.254.1.1 dev bpfin.cali 
```

**Comment by tomastigera** [comment]
No, just UDP services when `BPFConnectTimeLoadBalancing` is set to `TCP` and for all if it is set to `Disabled` (iirc) we had a fix for MTU issues with `bpfin.cali` for 3.28.1 so if might be worth upgrading if you have not done so yet.  Are only the services with the static entries affected? https://github.com/projectcalico/calico/pull/8922

**Comment by diogenxs** [comment]
> just UDP services when BPFConnectTimeLoadBalancing is set to TCP

this is true, all static entries are services running with UDP, DNS servers basically 

> Are only the services with the static entries affected?

yes, only UDP, therefore the static entries

I'll upgrade and test again, currently running `3.27.4` :face_with_peeking_eye:

**Comment by tomastigera** [comment]
@nick-oconnor Finally making some progress on the fragmentation issue? Do I understand correctly, that your usecase is mainly intra-cluster UDP traffic?

**Comment by nick-oconnor** [comment]
Yep. I have a k8s workload that scrapes metrics from various pieces of networking hardware (outside the cluster) via SNMP.

**Comment by tomastigera** [comment]
@nick-oconnor i have a tentative fix for your fragmentation problem. Would you be able to test it out? What version of calico do you use?

**Comment by nick-oconnor** [comment]
Hi @tomastigera! Sorry for the delay, I was away for a few weeks. I'm on 3.30.2, operator version 1.38.3 if that matters. I see it's been merged, but I'm more than happy to test it out. Glancing at the operator values, I don't see a way to override the calico version. Need be I can scale it down and set image tags by hand.

**Comment by nick-oconnor** [comment]
Nvm, looks like I can use the ImageSet CRD. Nice design! 😎

**Comment by tomastigera** [comment]
I think the easiest with operator is to set `unsupported.operator.tigera.io/ignore: "true"` on the calico-node ds https://github.com/tigera/operator/blob/master/docs/common_tasks.md#making-temporary-changes-to-components-the-operator-manages and set `image` to `calico/node:v3.31.0-0.dev-494-ge271c227c865` which has the fix. It should work well with 3.30.2 base.

**Comment by elsoa-invitech** [comment]
How can i enable this?
I have new calico: 

    image: quay.io/calico/cni:v3.31.2
    imageID: quay.io/calico/cni@sha256:62f50ed3070fd5da4df907fc2d618faf53f2d69d2c64232f7f464a646806cb5b

But large ping is not working:

```
test1-cch8p:~# ping -c3 -W 1 -s 1400 10.243.177.131
PING 10.243.177.131 (10.243.177.131) 1400(1428) bytes of data.
1408 bytes from 10.243.177.131: icmp_seq=1 ttl=63 time=0.036 ms
1408 bytes from 10.243.177.131: icmp_seq=2 ttl=63 time=0.038 ms
1408 bytes from 10.243.177.131: icmp_seq=3 ttl=63 time=0.126 ms

--- 10.243.177.131 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2034ms
rtt min/avg/max/mdev = 0.036/0.066/0.126/0.041 ms
test1-cch8p:~# ping -c3 -W 1 -s 2400 10.243.177.131
PING 10.243.177.131 (10.243.177.131) 2400(2428) bytes of data.

--- 10.243.177.131 ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 2032ms
```
